### PR TITLE
Use a 8 bit-flag to store all avalonia property flags

### PR DESCRIPTION
**Pull request by heku**
## What does the pull request do?
To reduce the memory cost by combining four flag fields into one, similar as [WPF's](https://referencesource.microsoft.com/#WindowsBase/Base/System/Windows/DependencyProperty.cs,1e729c9b0868ea96).

## What is the current behavior?
Every AvaloniaProperty needs four fields to store the flags.
```cs
public bool Inherits { get; private protected set; }
public bool IsAttached { get; private protected set; }
public bool IsDirect { get; private protected set; }
public bool IsReadOnly { get; private protected set; }
```

## How was the solution implemented (if it's not obvious)?
```cs
[Flags]
private enum Flag : byte
{
    Inherits = 1,   // 0b_0000_0001
    IsAttached = 2, // 0b_0000_0010
    IsDirect = 4,   // 0b_0000_0100
    IsReadOnly = 8  // 0b_0000_1000
}
private Flag _flags;
public bool Inherits
{
    get => _flags.HasFlag(Flag.Inherits);
    private protected set => _flags = value ? (_flags | Flag.Inherits) : (_flags & ~Flag.Inherits);
}
public bool IsAttached
{
    get => _flags.HasFlag(Flag.IsAttached);
    private protected set => _flags = value ? (_flags | Flag.IsAttached) : (_flags & ~Flag.IsAttached);
}
public bool IsDirect
{
    get => _flags.HasFlag(Flag.IsDirect);
    private protected set => _flags = value ? (_flags | Flag.IsDirect) : (_flags & ~Flag.IsDirect);
}
public bool IsReadOnly
{
    get => _flags.HasFlag(Flag.IsReadOnly);
    private protected set => _flags = value ? (_flags | Flag.IsReadOnly) : (_flags & ~Flag.IsReadOnly);
}
```

### CODE REVIEW

**Comment by Gillibald** [comment]
I would prefer a bitwise check instead of HasFlag for performance reasons

**Comment by heku** [comment]
Make sense, I've updated it accordingly.

**Comment by danipen** [comment]
Just out of curiosity. What is the memory gain? 3 bytes per object right? If so, it's aprox 3Mb each 1million objects... Or am I missing something due to memory alignment or something else?

**Comment by timunie** [comment]
I'd be interested in a performance comparison here as well. Would be interesting to see what the impact is.

**Comment by hez2010** [comment]
> I would prefer a bitwise check instead of HasFlag for performance reasons

This makes no difference as HasFlag can already be highly optimized by the JIT:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKBoDEAbbAcwAIB9AMyedwG4aNYBAgNWASQB2ACxhQAlhlwAKAJQ0A3jVY7WxAOwduLXADoAEnkYtl15qamyFS1QOoBfQdQDad3AF0aGEkAV3xWO1YQVmAATwwYTW1dRzlFXFYAXlZyNF0AenzWAAZgdmKK4vKKqmpdCVwAQQwMbDBZABMs1lI8wpKyyqqa4uSdcVwAEXlYMAxu9AKi0uqK8vIKsYaAJRhsDoB5SQZY7oAOHX6VofYNzY8gA=

**Comment by Gillibald** [comment]
@hez2010 Are you sure this is the case for all runtimes?

You can also use this: https://github.com/YohDeadfall/Avalonia/blob/master/src/Avalonia.Base/EnumExtensions.cs

**Comment by heku** [comment]
I did some tests, the result turns out there's no memory difference so far, because of the memory aligment.

```cs
public class MockObject : AvaloniaObject
{
    public readonly AvaloniaProperty[] array = new AvaloniaProperty[30000];

    public void Setup()
    {
        for (int i = 0; i < 10000; i += 3)
        {
            array[i]     = AvaloniaProperty.Register<MockObject, string>("p" + i);
            array[i + 1] = AvaloniaProperty.RegisterAttached<MockObject, MockObject, string>("a" + i);
            array[i + 2] = AvaloniaProperty.RegisterDirect<MockObject, string>("d" + i, x => null);
        }
    }
}

var o = new MockObject();

GC.Collect();
GC.WaitForPendingFinalizers();
GC.Collect();

var bytes1 = GC.GetTotalMemory(true);
Console.WriteLine("{0} bytes", bytes1);

o.Setup();

GC.Collect();
GC.WaitForPendingFinalizers();
GC.Collect();

var bytes2 = GC.GetTotalMemory(true);
Console.WriteLine("{0} bytes, increase {1} Kb, {2} Mb", bytes2, (bytes2 - bytes1) / 1024.0, (bytes2 - bytes1) / 1024.0 / 1024.0);
Console.WriteLine("{0} bytes per instance", (bytes2 - bytes1) / (o.array.Length * 1.0));
GC.KeepAlive(o);
```

Both print on my machine (win11 x64)
```
329760 bytes
7547208 bytes, increase 7048.2890625 Kb, 6.883094787597656 Mb
240.5816 bytes per instance
```
But if one more flag added, you'll see the difference, bitwise takes a little less memory than boolean fields.

So, please feel free to close the PR.

**Comment by kekekeks** [comment]
You can see the memory layout using this tool:

https://github.com/SergeyTeplyakov/ObjectLayoutInspector

It might be worth to manually rearrange fields for better memory layout.

**Comment by maxkatz6** [comment]
Either way, there shouldn't be many AvaloniaProperty objects in memory. Similar optimizations might make more sense for Visuals though. Or styled property values, but that part was already optimized this way iirc.

File: src/Avalonia.Base/AvaloniaProperty.cs
Comment (damian-666) [inline src/Avalonia.Base/AvaloniaProperty.cs:1]: jsut another eyes on this please   .. looks like the bits could  be combined before but now not sure to be.  or use extension methods.. the logic looks off to me but i 

 set => _flags **=** value ? (_flags | Flag.Inherits) : (_flags & ~Flag.Inherits);   doesnt this notify even if not changed.. what does assingment do in condtion?   ==?  i dont know super c# nuances but something looks off.
 my guess
  **__set =>   value == true? _falgs &=!Flag.Inherits :  _flags & ~Flag.Inherits  ???      also 32 bit in normal for enums and is in WPF , due to alignment but maybe the JITTE can pack the bytes but other enums are nomral 32 bit.__**
  
  [[Flags](https://referencesource.microsoft.com/mscorlib/A.html#5bbb5168ad9d1021)]
        private enum [Flags](https://referencesource.microsoft.com/WindowsBase/R/1e729c9b0868ea96.html) **: int**
        {
            [GlobalIndexMask](https://referencesource.microsoft.com/WindowsBase/R/ff382ec690109e06.html)                           = 0x0000FFFF,
            [IsValueType](                               = 0x00010000,
            [IsFreezableType]                          = 0x00020000,
            [IsStringType](https://referencesource.microsoft.com/WindowsBase/R/26b1dd1b0d584a67.html)                              = 0x00040000,
            [IsPotentiallyInherited](https://referencesource.microsoft.com/WindowsBase/R/5968e711840222c4.html)                    = 0x00080000,
            [IsDefaultValueChanged](https://referencesource.microsoft.com/WindowsBase/R/6e4bf9e980ce98fc.html)                     = 0x00100000,
            [IsPotentiallyUsingDefaultValueFactory](https://referencesource.microsoft.com/WindowsBase/R/24d2102347a4c848.html)     = 0x00200000,
            [IsObjectType](https://referencesource.microsoft.com/WindowsBase/R/cdf809a7d8f761f1.html)                              = 0x00400000,
            // 0xFF800000   free bits
        }

**Pull request closed without merging**